pkg/volume/csi/csi_mounter.go

```diff
@@ -134,68 +150,119 @@ func (c *csiMountMgr) SetUpAt(dir string, fsGroup *int64) error {
 			return err
 		}
 	}
-	// search for attachment by VolumeAttachment.Spec.Source.PersistentVolumeName
-	if c.volumeInfo == nil {
-		nodeName := string(c.plugin.host.GetNodeName())
-		c.volumeInfo, err = c.plugin.getPublishVolumeInfo(c.k8s, c.volumeID, c.driverName, nodeName)
+
+	// if source is an inline CSIVolumeSource:
+	//   - if needed, provision
+	//   - then attach
+	//   - then continue below for remaining mount steps
+	// else check for SetUp of normal CSIPersistentVolumeSource
+	if c.plugin.inlineFeatureEnabled && volSrc != nil {
+		klog.V(4).Info(log("mounter.SetUpAt inline from CSIVolumeSource"))
+
+		if volSrc.VolumeHandle != nil {
+			volumeHandle = *volSrc.VolumeHandle
+		} else {
+			klog.Error(log("mounter.SetUpAt failed as VolumeHandle is nil", driverName))
+		}
+
+		if volSrc.FSType != nil {
+			fsType = *volSrc.FSType
+		}
+
+		volAttribs = volSrc.VolumeAttributes
+		klog.V(4).Info(log("mounter.SetUpAt getting PublishVolumeInfo"))
+		volumeInfo, err = c.plugin.getPublishVolumeInfo(c.k8s, volumeHandle, driverName, nodeName)
 		if err != nil {
+			klog.V(4).Info(log("mounter.SetUpAt failed to get VolumeInfo: %v", err))
 			return err
 		}
-	}
-
-	attribs := csiSource.VolumeAttributes
 
-	nodePublishSecrets := map[string]string{}
-	if csiSource.NodePublishSecretRef != nil {
-		nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, csiSource.NodePublishSecretRef)
+		if volSrc.NodePublishSecretRef != nil {
+			secretName := volSrc.NodePublishSecretRef.Name
+			ns := c.pod.Namespace
+			nodePublishSecrets, err = getCredentialsFromSecret(c.k8s, &api.SecretReference{Name: secretName, Namespace: ns})
+			if err != nil {
+				return fmt.Errorf("fetching NodePublishSecretRef %s/%s failed: %v", ns, secretName, err)
+			}
+		}
+		mountOptions = []string{}
+
+		klog.V(4).Info(log("mounter.SetUpAt inline from CSIVolumeSource OK [driver:%d, volumeHandle:%s]", driverName, volumeHandle))
```

Comment:
this log isn't accurate. I wouldn't declare `SetUpAt` succesful until after the Publish call succeeds